src/HealthGPS/energy_balance_model.cpp

```diff
@@ -35,66 +46,163 @@ void EnergyBalanceModel::generate_risk_factors([[maybe_unused]] RuntimeContext &
 }
 
 void EnergyBalanceModel::update_risk_factors(RuntimeContext &context) {
-    auto age_key = core::Identifier{"age"};
-    for (auto &entity : context.population()) {
-        // Ignore if inactive, newborn risk factors must be generated, not updated!
-        if (!entity.is_active() || entity.age == 0) {
+    hgps::Population &population = context.population();
+    const size_t population_size = population.current_active_size();
+    double mean_sim_body_weight = 0.0;
+    double mean_adjustment_coefficient = 0.0;
+
+    // TODO: Compute target body weight.
+    const float target_BW = 100.0;
+
+    // Trial run.
+    for (auto &person : population) {
+        // Ignore if inactive.
+        if (!person.is_active()) {
             continue;
         }
 
-        auto current_risk_factors =
-            get_current_risk_factors(context.mapping(), entity, context.time_now());
-
-        // Model calibrated on previous year's age
-        auto model_age = static_cast<int>(entity.age - 1);
-        if (current_risk_factors.at(age_key) > model_age) {
-            current_risk_factors.at(age_key) = model_age;
-        }
+        auto trial = simulate_person(person, false, 0.0);
+        mean_sim_body_weight += trial[0];
+        mean_adjustment_coefficient += trial[1];
+    }
 
-        [[maybe_unused]] double energy_intake = 0.0;
-        std::unordered_map<core::Identifier, double> nutrient_intakes;
+    // Compute baseline adjustment.
+    mean_sim_body_weight /= population_size;
+    mean_adjustment_coefficient /= population_size;
+    double final_shift = (target_BW - mean_sim_body_weight) / mean_adjustment_coefficient;
 
-        // Initialise nutrients to zero.
-        for (const auto &coefficient : energy_equation_) {
-            nutrient_intakes[coefficient.first] = 0.0;
+    // Final run.
+    for (auto &person : population) {
+        // Ignore if inactive.
+        if (!person.is_active()) {
+            continue;
         }
 
-        // Compute nutrient intakes from food intakes.
-        for (const auto &equation : nutrient_equations_) {
-            // TODO: until carbs is added to risk factors (should use at, not [])
-            // double food_intake = current_risk_factors.at(equation.first);
-            double food_intake = current_risk_factors[equation.first];
+        simulate_person(person, true, final_shift);
+    }
+}
 
-            for (const auto &coefficient : equation.second) {
-                double delta_nutrient = food_intake * coefficient.second;
-                nutrient_intakes[coefficient.first] += delta_nutrient;
-            }
-        }
+std::array<double, 2> EnergyBalanceModel::simulate_person(Person &person, bool final_run,
```

Comment:
I feel like this should return a `std::pair<double, double>` because the values are unrelated.

Reply:
I'm quite suspicious about this `final_run` argument; it kind of feels like you're trying to squash two functions into one here (with two different return types).

Would it make sense to just return the processed values (H, BW, etc.) in a struct? It could then be fed into either a function which updates the `Person` or calculates the trial adjustment depending on what the user wants to do with it. What do you reckon?

Reply:
Fair enough. The baseline adjustment would be computed unnecessarily again on the final run, but that's not such a huge deal.

Reply:
Is there a way the baseline adjustment could be calculated in its own function?

Maybe once you've split this function up a bit, the way forward will seem more obvious.

Reply:
I've thought about it, but it's not obvious how to do this. It relies on mean simulated weights and mean of these adjustment coefficients for the whole population, so I'd have to sit down and do a bit of math to be able to decouple it.